Path in error message is neither public name nor internal name of type

#### dep/src/lib.rs

```rust
pub struct S;
mod m { pub struct S; }
pub use crate::m::S as S2;
```

#### repro/src/lib.rs

```rust
extern crate dep;
fn f() {
    let _: dep::S = dep::S2;
}
```

The error message refers to both types as `dep::S` even though one is `dep::S2`. I am reporting this as distinct from #21934 because `dep::S` isn't even the internal name of the other struct. That would be `dep::m::S`. Printing `dep::S2` would be ideal (as tracked in #21934) but printing `dep::m::S` would still be a big improvement.

```console
error[E0308]: mismatched types
 --> src/lib.rs:4:21
  |
4 |     let _: dep::S = dep::S2;
  |                     ^^^^^^^ expected struct `dep::S`, found a different struct `dep::S`
  |
  = note: expected type `dep::S` (struct `dep::S`)
             found type `dep::S` (struct `dep::S`)
```

#### Same thing in script form:

```bash
#!/bin/bash

cargo new --lib dep
cargo new --lib repro

echo >dep/src/lib.rs '
pub struct S;
mod m { pub struct S; }
pub use crate::m::S as S2;
'

echo >repro/src/lib.rs '
extern crate dep;
fn f() {
    let _: dep::S = dep::S2;
}
'

echo >>repro/Cargo.toml 'dep = { path = "../dep" }'

# rustc 1.33.0-nightly (adbfec229 2018-12-17)
cargo +nightly check --manifest-path repro/Cargo.toml
```

<br>

Mentioning @dgreid who hit this today. The real case involves multiple `Error` types across different modules, each re-exported from lib.rs under a unique name.

--- a/src/librustc/ty/item_path.rs
+++ b/src/librustc/ty/item_path.rs
@@ -210,12 +210,12 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
 
             let visible_parent = visible_parent_map.get(&cur_def).cloned();
             let actual_parent = self.parent(cur_def);
-            debug!(
-                "try_push_visible_item_path: visible_parent={:?} actual_parent={:?}",
-                visible_parent, actual_parent,
-            );
 
             let data = cur_def_key.disambiguated_data.data;
+            debug!(
+                "try_push_visible_item_path: data={:?} visible_parent={:?} actual_parent={:?}",
+                data, visible_parent, actual_parent,
+            );
             let symbol = match data {
                 // In order to output a path that could actually be imported (valid and visible),
                 // we need to handle re-exports correctly.
@@ -248,16 +248,16 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
                 // the children of the visible parent (as was done when computing
                 // `visible_parent_map`), looking for the specific child we currently have and then
                 // have access to the re-exported name.
-                DefPathData::Module(module_name) if visible_parent != actual_parent => {
-                    let mut name: Option<ast::Ident> = None;
-                    if let Some(visible_parent) = visible_parent {
-                        for child in self.item_children(visible_parent).iter() {
-                            if child.def.def_id() == cur_def {
-                                name = Some(child.ident);
-                            }
-                        }
-                    }
-                    name.map(|n| n.as_str()).unwrap_or(module_name.as_str())
+                DefPathData::Module(actual_name) |
+                DefPathData::TypeNs(actual_name) if visible_parent != actual_parent => {
+                    visible_parent
+                        .and_then(|parent| {
+                            self.item_children(parent)
+                                .iter()
+                                .find(|child| child.def.def_id() == cur_def)
+                                .map(|child| child.ident.as_str())
+                        })
+                        .unwrap_or_else(|| actual_name.as_str())
                 },
                 _ => {
                     data.get_opt_name().map(|n| n.as_str()).unwrap_or_else(|| {

--- a/src/test/ui/issues/auxiliary/issue-56943.rs
+++ b/src/test/ui/issues/auxiliary/issue-56943.rs
@@ -1,0 +1,3 @@
+pub struct S;
+mod m { pub struct S; }
+pub use crate::m::S as S2;

--- a/src/test/ui/issues/issue-56943.rs
+++ b/src/test/ui/issues/issue-56943.rs
@@ -1,0 +1,8 @@
+// aux-build:issue-56943.rs
+
+extern crate issue_56943;
+
+fn main() {
+    let _: issue_56943::S = issue_56943::S2;
+    //~^ ERROR mismatched types [E0308]
+}

--- a/src/test/ui/issues/issue-56943.stderr
+++ b/src/test/ui/issues/issue-56943.stderr
@@ -1,0 +1,12 @@
+error[E0308]: mismatched types
+  --> $DIR/issue-56943.rs:6:29
+   |
+LL |     let _: issue_56943::S = issue_56943::S2;
+   |                             ^^^^^^^^^^^^^^^ expected struct `issue_56943::S`, found struct `issue_56943::S2`
+   |
+   = note: expected type `issue_56943::S`
+              found type `issue_56943::S2`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

